images/archive/importer.go

```diff
@@ -186,15 +187,24 @@ func ImportIndex(ctx context.Context, store content.Store, reader io.Reader, opt
 			return ocispec.Descriptor{}, errors.Wrap(err, "write docker manifest")
 		}
 
-		platforms, err := images.Platforms(ctx, store, desc)
+		imgPlatforms, err := images.Platforms(ctx, store, desc)
 		if err != nil {
 			return ocispec.Descriptor{}, errors.Wrap(err, "unable to resolve platform")
 		}
-		if len(platforms) > 0 {
+
+		if len(imgPlatforms) == 0 || (imgPlatforms[0].OS == "windows" && imgPlatforms[0].OSVersion == "") {
```

Comment:
`len(imgPlatforms) == 0` makes sense. If the image itself cannot tell the platform, we can fallback to the host's platform.

`(imgPlatforms[0].OS == "windows" && imgPlatforms[0].OSVersion == "")` still confuses me. The image explicitly says that it is for Windows, why does it fallback to Linux?

Reply:
This is attempting to solve the problem of a windows image that does not have the extra information (the `OSVersion`) that is effectively required by the Windows runtime layer.

As you note, it introduces a potential problem: what if you import an `OS == windows` image on a daemon not on the Windows platform? It would set the platform content to completely incorrect values.

I don't know if there is an easy solution here; @claudiubelu's assumption (I think?) is that most likely the image will be usable on the version collected from the Windows host. Do we need to add a warning or error out if we are trying to come up with a Windows `OSVersion` but `runtime.GOOS` is not Windows?

Reply:
> 
> `(imgPlatforms[0].OS == "windows" && imgPlatforms[0].OSVersion == "")` still confuses me. The image explicitly says that it is for Windows, why does this code fallback to Linux or other operating systems?

The ``platforms.DefaultSpec()`` for Windows is this: [1]. As you can see, the OS value is ``runtime.GOOS``, which on Windows will be ``windows``, so we'll still say that the Windows image is a Windows image. We're basically just adding the ``OSVersion`` to it.


> 
> 
> This is attempting to solve the problem of a windows image that does not have the extra information (the `OSVersion`) that is effectively required by the Windows runtime layer.

Right, we need that ``OSVersion`` information. The fact that it's missing from the Image spec [1] is problematic. Ideally, that field would be present there, so we can actually detect that a user is trying to import an image with an older OS Version that is not supported on the current OS. But I don't think the ``DockerV2.0`` and ``DockerV2.1`` images will have that field either way.

> 
> As you note, it introduces a potential problem: what if you import an `OS == windows` image on a daemon not on the Windows platform? It would set the platform content to completely incorrect values.

Oh I see, that makes sense. Will address this issue shortly.

> 
> I don't know if there is an easy solution here; @claudiubelu's assumption (I think?) is that most likely the image will be usable on the version collected from the Windows host. Do we need to add a warning or error out if we are trying to come up with a Windows `OSVersion` but `runtime.GOOS` is not Windows?

Right. Ideally, we'd update the image spec [2], but I don't think that will help with the current I don't think the ``DockerV2.0`` and ``DockerV2.1`` images. Right now, we need that ``OSVersion`` field in order to pass the ``FilterPlatforms`` handler in the image import [3], without which the image's content won't have any children content, which means it will eventually get Garbage Collected, breaking the image (even though ctr import's exit code was 0).

Right now, the only Windows images that can be correctly imported through ctr are images that have only 1 layer (since they don't need any children labeled). This will solve that issue.

Additionally, the Garbage Collector deleting content from multi-layered images is not a Windows-specific issue. I've sent a PR here which proves the same issue on Linux as well: https://github.com/containerd/containerd/pull/5933

[1] https://github.com/containerd/containerd/blob/ff75f7d4899cdaa019be4cd904fc5c4ab9748b09/platforms/defaults_windows.go#L31
[2] https://github.com/containerd/containerd/blob/af1a0908d06b4f7db4a11bfc67ecc2f34baeb04b/vendor/github.com/opencontainers/image-spec/specs-go/v1/config.go#L82
[3] https://github.com/containerd/containerd/blob/af1a0908d06b4f7db4a11bfc67ecc2f34baeb04b/import.go#L169

Reply:
> Additionally, the Garbage Collector deleting content from multi-layered images is not a Windows-specific issue.

This only solves the Windows issue not Linux, correct? I don't see change in behavior for the non-windows scenario.

Reply:
Yes, for ``Docker`` formatted images, at least. There are 3 cases:

- the image platform is Windows, in which case we add the ``OSVersion`` if it's missing, allowing it to pass the platform matcher used by ``FilterPlatforms``.
- the image platform is Linux. We don't do anything in this case, since there is nothing to do. The platform matcher should pass, as long as it properly matches the OS platform [1] (so, ``linux/amd64`` images for ``linux/amd64`` hosts, etc.)
- the image platform is not set. The ``FilterPlatforms`` filter checks if a particular descriptor has the ``Platform`` set or not [2]. So, if it doesn't have any ``Platform`` set, then that descriptor is not filtered out, which is the same behaviour as an image matching the host plaform.

There are also ``OCI`` formatted images (btw, ctr exports ``OCI`` images). The situation there is trickier, and it can be seen that OCI images can still be garbage collected, even on Linux, as shown here [3]. The issue there is different, it's not because of missing platforms; I think it's because ``images.Walk`` doesn't walk over all the image content, which is required in order to add the child labels, which means that that content will get garbage collected. I haven't solved this issue.

[1] https://github.com/containerd/containerd/blob/d0bedc5bd111172f5a14fcbf83ad19ab471337c5/platforms/defaults_unix.go#L40
[2] https://github.com/containerd/containerd/blob/d0bedc5bd111172f5a14fcbf83ad19ab471337c5/images/handlers.go#L240
[3] https://github.com/containerd/containerd/pull/5933